syft/messaging/plan.py

```diff
@@ -817,9 +858,53 @@ def __str__(self):
 
         return out
 
+    @property
+    def output_shape(self):
+
+        if self._output_shape is None:
+
+            args = [th.zeros(shape) for shape in self.input_shapes]
+            output = self(*args)
+            self._output_shape = output.shape
+
+        return self._output_shape
+
     def __repr__(self):
         return self.__str__()
 
+    def setup_plan_with_promises(self, *args):
+        """ Slightly modifies a plan so that it can work with promises.
+        The plan will also be sent to location with this method.
+        """
+        location = args[0].child.location if isinstance(args[0].child, PointerTensor) else None
+        if location is not None:
+            # Get remote arguments
+            args = [arg.child.location[arg.child.id_at_location] for arg in args]
+
+        prom = None
+        for p in args:
+            if hasattr(p, "child") and isinstance(p.child, PromiseTensor):
+                p.child.plans.add(self.id)
+                if prom is None:
+                    prom = p
+
+        res = sy.PromiseTensor(
+            owner=prom.owner,
+            shape=self.output_shape,
+            tensor_id=self.result_ids[0],
+            tensor_type=prom.child.obj_type,
```

Comment:
not sure you can infer the type of the res from the input type: maybe just let it empty for now and replaced by a default "th.floattensor"